README.md
<!--
 * @Author: ShawnPhang
 * @Date: 2021-09-02 10:44:23
 * @Description: 前端组件库搭建
 * @LastEditors: ShawnPhang <site: book.palxp.com>
 * @LastEditTime: 2023-07-20 16:37:14
 * @site: book.palxp.com / blog.palxp.com
-->

## 前端军火库

<!-- START -->

#### [ColorPicker 颜色选择器（吸色器）](/#/color-picker)

#### [ScrollWrap 吸附滚动组件](/#/scroll-wrap)

#### <a target="_blank" href="/#/j-editor">JEditor 掘金 MarkDown 编辑器</a>

#### [CodePanel 代码编辑器](/#/code-panel)

#### [IJS Utils 工具函数集](/#/ijs-utils)

<!-- END -->

## 说明

> 这是一个用于开发 Vue3 组件/组件库、以及工具类函数的仓库。
>
> 自动化的文档网站，只需专注于组件的开发，文档全自动生成。

## 为什么组件化

> 拆分组件的意义在于，细化项目的颗粒度，避免将复杂功能耦合在项目中，更方便后续的迁移改造、重写或移除、性能调优等工作。
>
> 在此基础上，高度抽象的组件还可复用，为其它项目中快速集成功能提供了可能性。

## 流程

1. 运行 `npm run add` 添加组件

2. 在 `packages` 中开发代码，在 `src/pages` 下编写测试/示例页面。（代码展示面板部分暂无热更新，可手动运行 `npm run vue-to-md` 来即时更新。）

3. 运行 `npm run change` 生成组件更新日志（lerna）

4. 运行 `npm run build:pkg` 编译组件

5. 运行 `npm run publish` 发布组件

6. 运行 `npm run build` 然后提交更新文档网站。（可能需要一些时间，因为会将所有组件案例打包）

## 设计

独立打包只生成一个 index 包，多文件打包（即按需打包）则是否存在 src 文件夹判定，分包在 src 开发。

组件工厂：packages/

组件目录：packages/xxxx/

组件打包文件夹：dist
组件打包 d.ts 文件夹：typings

## 基础

模块包管理：lerna

组件设计：按需加载

单元测试：mocha、jest

持续集成、代码规范、提交规范

## TODOLIST

1. [x] 按需打包，独立打包
2. [x] 指定打包
3. [ ] 对 d.ts 声明文件进行打包
4. [x] 提交组件的测试，lerna
5. [x] 文档的自动生成

## 视图组件库

区分：Vue2 Vue3 基于 UI 框架

打包：rollup

组件库文档：自动生成

## 函数类工具库

打包：rollup

文档工具：JSDoc

## 组件库按需加载示例（vite）

npm install vite-plugin-importer --save

```
yarn add vite-plugin-importer
```

```
// vite.config.js
import { defineConfig } from "vite";
import usePluginImport from 'vite-plugin-importer'
export default defineConfig({
  plugins: [
    ... // other plugins
    usePluginImport({
      libraryName: "ant-design-vue",
      libraryDirectory: "es",
      style: "css",
    }),
    usePluginImport({
      libraryName: "antd",
      libraryDirectory: "es",
      style: "css",
    }),
    usePluginImport({
      libraryName: 'vant',
      libraryDirectory: 'es',
      style: (name) => `${name}/style/less`,
    }),
    usePluginImport({
      libraryName: 'element-plus',
      customStyleName: (name) => {
        return `element-plus/lib/theme-chalk/${name}.css`;
      },
    })
    // Other configurations welcome PR
  ];
})
```

## 组件库按需加载示例（webpack）

```
yarn add babel-plugin-import -D
# 或
npm install babel-plugin-import -D
```

```
// babel.config.js
module.exports = {
  plugins: [
    [
      'import',
      {
        libraryName: 'element-plus',
        // 引入组件
        customName: (name) => {
          return `element-plus/lib/components/${name}`
        },
        // 引入样式
        // customStyleName: (name) => {
        //   return `element-plus/lib/components/${name}/style/css`
        // },
      },
    ],
  ],
}
```

## 打包发布

查看当前包源

```
nrm ls
```

切换源

```
nrm use npm
```

查看登录用户

```
npm whoami
```

没有则登录

```
npm login
# OR
npm addUser
```

## 备注

项目启动时 Node 环境：14.17.0 当前使用 Node 版本：v16.18.1


CHANGELOG.md
## [0.0.2](https://github.com/palxiao/front-end-arsenal/compare/v0.0.1...v0.0.2) (2023-09-05)

### 🌟 新功能
范围|描述|commitId
--|--|--
 升级 | 添加代码美化及校验等 | [817b90a](https://github.com/palxiao/front-end-arsenal/commit/817b90a)
 none | 新增掘金md编辑器 | [6582682](https://github.com/palxiao/front-end-arsenal/commit/6582682)
 none | 修改中 | [c5bdbfb](https://github.com/palxiao/front-end-arsenal/commit/c5bdbfb)
 component | 新增 ColorPicker 组件 | [effebc9](https://github.com/palxiao/front-end-arsenal/commit/effebc9)
 component | color-picker add blur | [1a3d1b0](https://github.com/palxiao/front-end-arsenal/commit/1a3d1b0)
 custom | vue-docs | [60a0975](https://github.com/palxiao/front-end-arsenal/commit/60a0975)
 example | 增加了代码逐行高亮逻辑 | [142a167](https://github.com/palxiao/front-end-arsenal/commit/142a167)
 package | 新增工具函数包 | [6d764d4](https://github.com/palxiao/front-end-arsenal/commit/6d764d4)
 script | 添加编译选择器 | [49f66a0](https://github.com/palxiao/front-end-arsenal/commit/49f66a0)
 none | update | [7844d14](https://github.com/palxiao/front-end-arsenal/commit/7844d14)
 none | version change | [bc2a86f](https://github.com/palxiao/front-end-arsenal/commit/bc2a86f)


### 🏠 构建系统
范围|描述|commitId
--|--|--
 none | 重新打包 | [5578342](https://github.com/palxiao/front-end-arsenal/commit/5578342)
 none | test | [16e4252](https://github.com/palxiao/front-end-arsenal/commit/16e4252)
 none | test publish | [fbc62fb](https://github.com/palxiao/front-end-arsenal/commit/fbc62fb)
 none | update | [397e3ed](https://github.com/palxiao/front-end-arsenal/commit/397e3ed)
 none | update | [e69a1e2](https://github.com/palxiao/front-end-arsenal/commit/e69a1e2)


### 🐛 Bug 修复
范围|描述|commitId
--|--|--
 none | 滚动监听 | [f335339](https://github.com/palxiao/front-end-arsenal/commit/f335339)
 none | 修复一个报错 | [41fdfc8](https://github.com/palxiao/front-end-arsenal/commit/41fdfc8)
 none | 修复auto docs watch bug | [f528f17](https://github.com/palxiao/front-end-arsenal/commit/f528f17)


### 📝 文档
范围|描述|commitId
--|--|--
 none | 更新文档 | [8d85898](https://github.com/palxiao/front-end-arsenal/commit/8d85898)
 none | 更新文档 | [362bb3e](https://github.com/palxiao/front-end-arsenal/commit/362bb3e)
 none | 增加一个文档 | [6300089](https://github.com/palxiao/front-end-arsenal/commit/6300089)
 all | 发布 | [e8e55c0](https://github.com/palxiao/front-end-arsenal/commit/e8e55c0)
 component | color选择器 | [aaef0d4](https://github.com/palxiao/front-end-arsenal/commit/aaef0d4)
 custom | readme update | [168e976](https://github.com/palxiao/front-end-arsenal/commit/168e976)
 data | package.json | [8e12889](https://github.com/palxiao/front-end-arsenal/commit/8e12889)
 docs | 更新文档 | [32120ee](https://github.com/palxiao/front-end-arsenal/commit/32120ee)
 none | readme | [4c012cd](https://github.com/palxiao/front-end-arsenal/commit/4c012cd)
 none | readme | [a17805e](https://github.com/palxiao/front-end-arsenal/commit/a17805e)
 none | readme | [848a963](https://github.com/palxiao/front-end-arsenal/commit/848a963)
 readme | 更新文档 | [494955d](https://github.com/palxiao/front-end-arsenal/commit/494955d)
 none | update | [9a10b33](https://github.com/palxiao/front-end-arsenal/commit/9a10b33)
 update | 更新vue-doc | [7585d62](https://github.com/palxiao/front-end-arsenal/commit/7585d62)


### 🔧 测试
范围|描述|commitId
--|--|--
 custom | test vue-docs | [8097c32](https://github.com/palxiao/front-end-arsenal/commit/8097c32)
 none | husky test | [ce2af2d](https://github.com/palxiao/front-end-arsenal/commit/ce2af2d)
 none | lerna publish | [f069d26](https://github.com/palxiao/front-end-arsenal/commit/f069d26)
 none | vue-doc | [abaf8e3](https://github.com/palxiao/front-end-arsenal/commit/abaf8e3)


### code
范围|描述|commitId
--|--|--
 none | 屏蔽docs的demo功能 | [5c8904f](https://github.com/palxiao/front-end-arsenal/commit/5c8904f)
 none | 优化 | [c635798](https://github.com/palxiao/front-end-arsenal/commit/c635798)
 none | 整理包 | [719c6b0](https://github.com/palxiao/front-end-arsenal/commit/719c6b0)
 component | 新增changelog | [28ca1c0](https://github.com/palxiao/front-end-arsenal/commit/28ca1c0)
 component | 优化code-panel | [0921c53](https://github.com/palxiao/front-end-arsenal/commit/0921c53)
 component | 暂时隐藏文档插件的代码高亮 | [41ea09a](https://github.com/palxiao/front-end-arsenal/commit/41ea09a)
 component | color-picker | [3f05e51](https://github.com/palxiao/front-end-arsenal/commit/3f05e51)
 component | update color-picker | [e65ac90](https://github.com/palxiao/front-end-arsenal/commit/e65ac90)
 custom | vue-docs 简化组件名 | [b59f2d3](https://github.com/palxiao/front-end-arsenal/commit/b59f2d3)
 none | test | [fd5536b](https://github.com/palxiao/front-end-arsenal/commit/fd5536b)
 none | vue-docs | [bee1736](https://github.com/palxiao/front-end-arsenal/commit/bee1736)
 none | vue-docs | [bdab32d](https://github.com/palxiao/front-end-arsenal/commit/bdab32d)
 none | vue-docs | [26ccd61](https://github.com/palxiao/front-end-arsenal/commit/26ccd61)


范围|描述|commitId
--|--|--
 none | Publish | [217d24c](https://github.com/palxiao/front-end-arsenal/commit/217d24c)
 none | Publish | [fd5d423](https://github.com/palxiao/front-end-arsenal/commit/fd5d423)
 none | Publish | [06cf966](https://github.com/palxiao/front-end-arsenal/commit/06cf966)
 none | Publish | [49730d6](https://github.com/palxiao/front-end-arsenal/commit/49730d6)
 none | Publish | [cd4ed16](https://github.com/palxiao/front-end-arsenal/commit/cd4ed16)
 none | Publish | [0a06a44](https://github.com/palxiao/front-end-arsenal/commit/0a06a44)
 none | Publish | [7369d1a](https://github.com/palxiao/front-end-arsenal/commit/7369d1a)
 none | Publish | [e27a5ad](https://github.com/palxiao/front-end-arsenal/commit/e27a5ad)
 none | Publish | [6c6017c](https://github.com/palxiao/front-end-arsenal/commit/6c6017c)
 none | Publish | [37f65bf](https://github.com/palxiao/front-end-arsenal/commit/37f65bf)
 none | Publish | [a3e5d06](https://github.com/palxiao/front-end-arsenal/commit/a3e5d06)
 none | Publish | [cda2cd2](https://github.com/palxiao/front-end-arsenal/commit/cda2cd2)

scripts/add.js
/*
 * @Author: ShawnPhang
 * @Date: 2023-05-25 17:45:07
 * @Description:
 * @LastEditors: ShawnPhang <site: book.palxp.com>
 * @LastEditTime: 2023-06-21 15:42:37
 */
const { exec } = require('child_process')
const chalk = require('chalk')
const inquirer = require('inquirer')
const ora = require('ora') // ora是一个命令行loading
const path = require('path')
const fs = require('fs')

const scope = '@palxp/' // 包的作用域，如果你的包没有作用域，那么则不需要

const packageJson = {
  name: '',
  version: '1.0.0',
  description: 'TODO',
  author: 'ShawnPhang <palxiao@vip.qq.com>',
  homepage: '',
  license: 'ISC',
  main: 'dist/index.js',
  module: 'dist/index.es.js',
  types: 'dist/index.d.ts',
  publishConfig: {
    access: 'public',
  },
  repository: {
    type: 'git',
    url: 'git+https://github.com/palxiao/front-end-arsenal.git',
  },
  scripts: {
    test: 'echo "Error: run tests from root" && exit 1',
  },
  bugs: {
    url: 'https://github.com/palxiao/front-end-arsenal/issues',
  },
}

const readme = (name) => {
  return `  <img style="display: inline-block;" src="https://img.shields.io/github/watchers/palxiao/front-end-arsenal?style=social" /> <img style="display: inline-block;" src="https://img.shields.io/github/forks/palxiao/front-end-arsenal?style=social" /> <img style="display: inline-block;" src="https://img.shields.io/github/stars/palxiao/front-end-arsenal?style=social" />
  
  # ${name}

  > TODO: 
  
  <img style="display: inline-block;" src="https://img.shields.io/npm/v/${scope}${name}" /> <img style="display: inline-block;" src="https://img.shields.io/bundlephobia/min/${scope}${name}?color=%2344cc88" /> <img style="display: inline-block;" src="https://img.shields.io/npm/dm/${scope}${name}" />

  ## Usage
  
  \`\`\`
  yarn add ${scope}${name}
  
  import ${name} from '${scope}${name}'
  \`\`\`
  
  ## API
  [API Docs 链接](/#/docs)
  <iframe src="/#/docs/${name}/index?preview=true" frameborder="0"></iframe>
  `
}

const index = () => {
  return `  import { App } from 'vue'
  import Comp from './index.vue'
  
  Comp.install = (app: App): void => {
    app.component(Comp.name, Comp)
  }
  
  export default Comp`
}

const pageVue = (name) => {
  return `<template>
  <example-frame :main-name="mainName" :readme-list="readmeList" :example-list="exampleList" />
</template>
<script lang="ts">
import ExampleFrame from '../../components/ExampleFrame.vue'
import { defineComponent } from 'vue'

export default defineComponent({
  components: { ExampleFrame },
  setup() {
    return {
      mainName: '${name}',
      readmeList: [
        {
          name: 'README',
          introduce: '说明文档',
        },{
          name: 'CHANGELOG',
          introduce: '更新日志',
        },
      ],
      exampleList: [
        {
          name: '基础演示',
          introduce: ''
        },
      ],
    }
  },
})
</script>`
}

const exampleVue = (name) => {
  return `<template><comp /></template>

  <script lang="ts">
  import comp from '@pkg/${name}/index.vue'
  import { defineComponent } from 'vue'
  export default defineComponent({
    components: { comp },
    setup() {
      return {}
    },
  })
  </script>`
}

inquirer
  .prompt([
    {
      type: 'input',
      name: 'name',
      message: '请输入组件名称',
      validate(input) {
        // 异步验证需要调用这个方法来告诉inquirer是否校验完成
        const done = this.async()
        input = String(input).trim()
        if (!input) {
          return done('请输入组件名称')
        }
        const spinner = ora('正在检查包名是否存在').start()
        exec(`npm search ${scope + input}`, (err, stdout) => {
          spinner.stop()
          if (err) {
            done('检查包名是否存在失败，请重试')
          } else {
            if (/No matches/.test(stdout)) {
              done(null, true)
            } else {
              done('该包名已存在，不允许创建!')
            }
          }
        })
      },
    },
  ])
  .then(async ({ name }) => {
    const fsFas = fs.existsSync(path.resolve(`./packages/${name}`))
    if (!fsFas) {
      fs.mkdir(path.resolve(`./packages/${name}`), { recursive: true }, (err) => {
        if (err) {
          console.error(err)
        } else {
          // 创建 packageJson
          packageJson.name = scope + name
          packageJson.homepage = `https://fe-doc.palxp.com/#/${name}`
          fs.writeFileSync(path.resolve(`./packages/${name}/package.json`), JSON.stringify(packageJson, null, '\t'))
          // 创建 README 文件
          fs.writeFileSync(path.resolve(`./packages/${name}/README.md`), readme(name))
          // 创建入口文件
          fs.writeFileSync(path.resolve(`./packages/${name}/index.ts`), index())
          // 创建网页
          fs.mkdirSync(path.resolve(`./src/pages/${name}`))
          fs.writeFileSync(path.resolve(`./src/pages/${name}/index.vue`), pageVue(name))
          fs.mkdirSync(path.resolve(`./src/pages/${name}/example`))
          fs.writeFileSync(path.resolve(`./src/pages/${name}/example/example1.vue`), exampleVue(name))

          console.log('创建成功，请访问：' + `http://localhost:3000/#/${name}`)
        }
      })
    } else {
      console.log('目录存在，包创建已终止!')
    }
    // const lernaDo = exec(`lerna create ${scope + name}`)
    // lernaDo.kill()
  })
  .catch((error) => {
    // 错误处理
    if (error.isTtyError) {
      console.log(chalk.bgRed.white.bold('当前环境无法运行，请手动添加组件'))
      process.exit(1)
    } else {
      console.log(chalk.bgRed.white.bold('未知错误，请手动添加组件'))
      process.exit(1)
    }
  })
